wallet/test/Test/Pos/Util/MnemonicsSpec.hs

```diff
@@ -0,0 +1,63 @@
+module Test.Pos.Util.MnemonicsSpec (spec, Entropy(..)) where
+
+import           Universum
+
+import           Data.ByteString.Char8 (pack)
+import           Data.Set (Set)
+import           Test.Hspec (Spec, describe, it, shouldSatisfy, xit)
+import           Test.Hspec.QuickCheck (modifyMaxSuccess)
+import           Test.QuickCheck (Arbitrary (..), property)
+import           Test.QuickCheck.Gen (Gen (..), oneof, vectorOf)
+import           Test.QuickCheck.Random (mkQCGen)
+
+import           Pos.Util.BackupPhrase (BackupPhrase (..), safeKeysFromPhrase)
+import           Pos.Util.Mnemonics (defMnemonic, fromMnemonic, toMnemonic)
+import           Pos.Wallet.Web.ClientTypes.Functions (encToCId)
+import           Pos.Wallet.Web.ClientTypes.Types (CId)
+
+import qualified Data.Set as Set
+
+
+spec :: Spec
+spec = do
+    describe "Pos.Util.Mnemonics" $ do
+        modifyMaxSuccess (const 10000) $ it "toMnemonic >=> fromMnemonic = Right" $ property $
+            \(Entropy ent) -> (toMnemonic ent >>= fromMnemonic) == Right ent
+
+        it "No example mnemonic" $
+            fromMnemonic defMnemonic `shouldSatisfy` isLeft
+
+        it "No empty mnemonic" $
+            (fromMnemonic "") `shouldSatisfy` isLeft
+
+        it "No empty entropy" $
+            (toMnemonic "") `shouldSatisfy` isLeft
+
+        xit "entropyToWalletId is injective " $
+            let
+                inject :: Ord b => (a -> b) -> [a] -> Set b
+                inject fn xs =
+                    foldl' (\acc x -> Set.insert (fn x) acc) Set.empty xs
+
+                genList :: Arbitrary a => Int -> [a]
+                genList n =
+                    unGen (vectorOf n arbitrary) (mkQCGen 14) 14
+
+                entropyToWalletId :: Entropy -> CId w
+                entropyToWalletId (Entropy ent) = cid
+                  where
+                    Right backupPhrase = (BackupPhrase . words) <$> toMnemonic ent
+                    Right cid          = (encToCId . fst) <$> safeKeysFromPhrase mempty backupPhrase
```

Comment:
These partial pattern matches are safe due to the invariant on the arbitrary instance for `Entropy`? I think I'd rather see an `either (error ("Invariant failed on Entropy in toMenomic: " <> BS.unpack ent)) id` to get a clearer error message for diagnostics; or even `either (error . show) id`

Reply:
They are safe indeed because the entropy we generate are valid (multiple of 4 bytes, given length, etc ...). Hence they _shouldn't fail_. You have a point however, if they ever do, message won't be much clearer.